File: driver_app/lib/services/location_service.dart
import 'package:flutter/foundation.dart';
import 'package:geolocator/geolocator.dart';

class LocationData {
  final double latitude;
  final double longitude;
  final double speed;
  final double heading;
  final int timestamp;
  final bool isStale;

  LocationData({
    required this.latitude,
    required this.longitude,
    required this.speed,
    required this.heading,
    required this.timestamp,
    required this.isStale,
  });
}

class LocationService {
  static final LocationService _instance = LocationService._internal();
  factory LocationService() => _instance;
  LocationService._internal();

  LocationData? _lastLocation;

  Future<bool> requestPermission() async {
    try {
      bool serviceEnabled = await Geolocator.isLocationServiceEnabled();
      if (!serviceEnabled) {
        return false;
      }

      LocationPermission permission = await Geolocator.checkPermission();
      if (permission == LocationPermission.denied) {
        permission = await Geolocator.requestPermission();
        if (permission == LocationPermission.denied) {
          return false;
        }
      }

      if (permission == LocationPermission.deniedForever) {
        return false;
      }

      return true;
    } catch (e) {
      debugPrint('[LocationService] Permission error: $e');
      return false;
    }
  }

  Future<LocationData?> getCurrentLocation() async {
    try {
      bool hasPermission = await requestPermission();
      if (!hasPermission) return null;

      Position pos = await Geolocator.getCurrentPosition(
        locationSettings: const LocationSettings(
          accuracy: LocationAccuracy.high,
          timeLimit: Duration(seconds: 4),
        ),
      );

      double speedKmh = pos.speed > 0 ? pos.speed * 3.6 : 0.0;
      double heading = pos.heading >= 0 ? pos.heading : 0.0;
      int now = DateTime.now().millisecondsSinceEpoch;

      final data = LocationData(
        latitude: pos.latitude,
        longitude: pos.longitude,
        speed: double.parse(speedKmh.toStringAsFixed(1)),
        heading: double.parse(heading.toStringAsFixed(1)),
        timestamp: pos.timestamp.millisecondsSinceEpoch,
        isStale: (now - pos.timestamp.millisecondsSinceEpoch) > 5000,
      );

      _lastLocation = data;
      return data;
    } catch (e) {
      debugPrint('[LocationService] getCurrentLocation error: $e');
      return null;
    }
  }

  LocationData? getLastLocation() => _lastLocation;
}

File: mobile/src/services/locationService.ts
import * as Location from 'expo-location';

export interface LocationData {
  latitude: number;
  longitude: number;
  speed: number; // km/h
  heading: number; // degrees
  timestamp: number; // ms epoch
  accuracy?: number | null;
  isStale: boolean;
}

class LocationService {
  private subscription: Location.LocationSubscription | null = null;
  private lastLocation: LocationData | null = null;

  async requestPermission(): Promise<boolean> {
    try {
      const { status } = await Location.requestForegroundPermissionsAsync();
      return status === 'granted';
    } catch (error) {
      console.warn('[LocationService] Permission error:', error);
      return false;
    }
  }

  async getCurrentLocation(): Promise<LocationData | null> {
    try {
      const hasPermission = await this.requestPermission();
      if (!hasPermission) return null;

      const location = await Location.getCurrentPositionAsync({
        accuracy: Location.Accuracy.High,
      });

      const speedKmh = location.coords.speed ? Math.max(0, location.coords.speed * 3.6) : 0;
      const heading = location.coords.heading && location.coords.heading >= 0 ? location.coords.heading : 0;
      const now = Date.now();

      const data: LocationData = {
        latitude: location.coords.latitude,
        longitude: location.coords.longitude,
        speed: Math.round(speedKmh * 10) / 10,
        heading: Math.round(heading * 10) / 10,
        timestamp: location.timestamp || now,
        accuracy: location.coords.accuracy,
        isStale: (now - (location.timestamp || now)) > 5000,
      };

      this.lastLocation = data;
      return data;
    } catch (error) {
      console.warn('[LocationService] Get location error:', error);
      return null;
    }
  }

  async startLocationUpdates(
    onLocationUpdate: (data: LocationData) => void,
    intervalMs: number = 2000
  ): Promise<boolean> {
    const hasPermission = await this.requestPermission();
    if (!hasPermission) return false;

    this.stopLocationUpdates();

    try {
      this.subscription = await Location.watchPositionAsync(
        {
          accuracy: Location.Accuracy.High,
          timeInterval: intervalMs,
          distanceInterval: 1, // 1 meter movement trigger
        },
        (location) => {
          const speedKmh = location.coords.speed ? Math.max(0, location.coords.speed * 3.6) : 0;
          const heading = location.coords.heading && location.coords.heading >= 0 ? location.coords.heading : 0;
          const now = Date.now();

          const data: LocationData = {
            latitude: location.coords.latitude,
            longitude: location.coords.longitude,
            speed: Math.round(speedKmh * 10) / 10,
            heading: Math.round(heading * 10) / 10,
            timestamp: location.timestamp || now,
            accuracy: location.coords.accuracy,
            isStale: (now - (location.timestamp || now)) > 5000,
          };

          this.lastLocation = data;
          onLocationUpdate(data);
        }
      );
      return true;
    } catch (error) {
      console.warn('[LocationService] Watch position error:', error);
      return false;
    }
  }

  stopLocationUpdates() {
    if (this.subscription) {
      this.subscription.remove();
      this.subscription = null;
    }
  }

  getLastLocation(): LocationData | null {
    if (!this.lastLocation) return null;
    const now = Date.now();
    return {
      ...this.lastLocation,
      isStale: (now - this.lastLocation.timestamp) > 5000,
    };
  }
}

export const locationService = new LocationService();
